F #he 948 improve calibration process

carebare47 opened the pull request:
## Proposed changes

Now runs faster / uses less compute

Fix progress bar logic

/tf callback was missing most messages due to callback execution time being too long. Now saves all messages and extracts a configurable subsample (can specify an approximate number of evenly distributed datapoints per second to keep)

Save result of previous sphere_fit as initial_guess for next sphere_fit (reduces computation time)

Add small sleep after tf subscriber, sometimes the subscriber warm-up time was longer than the first progress checkpoint, giving us empty lists which then got passed to np.std and threw numpy RuntimeWarnings

Added functionality to read from a rosbag for testing purposes 

Publishing a MarkerArray instead of lots of Markers (saves cpu time)


## Checklist

Before posting a PR ensure that from each of the below categories **AT LEAST ONE BOX HAS BEEN CHECKED**. If more than one category is applicable then more can be checked. Also ensure that the proposed changes have been filled out with relevant information for reviewers.

## Tests

- [ ] No tests required to be added. (For small changes that will be tested by CI/CD infrastructure).
- [ ] Added/Modified automated and PhantomHand CI tests (if a new class is added (Python or C++), the interface of that class must be unit tested).
- [ ] Manually tested in simulation (if simulation specific or no hardware required to test the functionality). 
- [ ] Manually tested on hardware (if hardware specific or related).

## Documentation

- [x] No documentation required to be added.
- [ ] Added documentation (For any new feature, explain what it does and how to use it. Write the documentation in a relevant space, e.g. Github, Confluence, etc).
- [ ] Updated documentation (For changes to pre-existing features mentioned in the documentation).

Commit:
fix progress bar logic

Commit:
tf subscriber takes time to start up. Added a small sleep to make sure we actually have data points when we first call fit_sphere(...) which prevents three numpy RuntimeWarnings (we were sometimes calling np.std on an empty list)

Commit:
add 'divisor' to tf callback, lets us only save every 'x' tf messages. Speeds up computation but potentially reduces accuracy. Needs testing

Commit:
save results from previous sphere_fit and use as the initial guess for the next sphere_fit. Speeds up computation time

Commit:
lint

Commit:
tidying

Commit:
removed unused import

bdenoun commented on src/polhemus_ros_driver/knuckle_position_calibration_action_server.py:
```diff
@@ -392,18 +399,20 @@ def _create_control(quaternion: Quaternion, name: str):
-                        data_point_marker = DataMarker(hand.polhemus_base_name, Point(*pos),
-                                                       COLORS[color_index].value)
-                        hand.pub.publish(data_point_marker)
+        if self._tf_data_counter % self._tf_data_divisor == 0:
```
```suggestion
        if not self._tf_data_counter % self._tf_data_divisor:
```

carebare47 replied:
== 0 is more explicit, we're checking for an integer value as the return from the modulo operator after all

bdenoun replied:
The conventions seem to disagree with that: https://pylint.readthedocs.io/en/latest/user_guide/messages/convention/use-implicit-booleaness-not-comparison-to-zero.html

carebare47 replied:
Well I disagree with the convention :stuck_out_tongue: don't really see how making the code less explicit helps anyone here, and it would make it a little less robust if any non integer types crept into `self._tf_data_counter` or `self._tf_data_divisor`

bdenoun commented on src/polhemus_ros_driver/knuckle_position_calibration_action_server.py:
```diff
@@ -140,6 +141,12 @@ def __init__(self, side: str = "right"):
         self._action_server = actionlib.SimpleActionServer("/calibration_action_server", CalibrateAction,
                                                            execute_cb=self._calibration, auto_start=False)
+        # How many times during calibration should we calculate the sphere fit and update quality %
+        self._number_of_checkpoints = 4
+        self._progress_period = 1.0 / self._number_of_checkpoints
+        # Only save every 'x'th glove data point for calibration - trade-off between accuracy and computation time
+        self._tf_data_divisor = 5
```
These seem to be constants, so I think they should be formatted as upper case. And these are values we potentially want to play with, so why not set a default value as init param (at least number of check point and data divisor)

carebare47 replied:
Good idea. Data divisor is something we'd want as an init param, but number of checkpoints should be changed to an internal constant

bdenoun replied:
Why internal constant? Maybe moving this value to 10 might make things better no? Or am I missing something?

bdenoun commented on src/polhemus_ros_driver/knuckle_position_calibration_action_server.py:
```diff
@@ -483,13 +495,23 @@ def _get_knuckle_positions(self, hand: Hand, plot=False):
             sphere_fit = SphereFit(data=hand.finger_data[finger]['data'], plot=plot)
 
+            initial_guess = None
+            if 'numpy' in str(type(hand.last_center_estimate)) and 'numpy' in str(type(hand.last_radius_estimate)):
```
Any reason why you don't use `isinstance(hand.last_center_estimate, np.ndarray)` or even `type(hand.last_center_estimate).__module__ == np.__name__`?

carebare47 replied:
Didn't trust that it would *always* be an ndarray. Didn't know about the second one :)

bdenoun replied:
How is `hand.last_center_estimate` computed?

carebare47 replied:
from [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.least_squares.html) and then split up [here](https://github.com/shadow-robot/polhemus_ros_driver/blob/e024ce5e698dc34c45afaf6b52d7b393f645db19/src/polhemus_ros_driver/sphere_fit.py#L124)

bdenoun replied:
So the center estimate is necessarily an ndarray, but the radius is likely to be an np int.

Commit:
add data divisor as input parameter (also raise exception on not-an-int)

Commit:
add data divisor as input parameter (also raise exception on not-an-int)

carebare47 commented:
found a bug - don't merge yet

Commit:
fix initial guess storage (was per hand, now per finger, now significantly faster), and fix parameter from launch file

Commit:
actually apply the fix I thought I'd apply in the previous commit

bdenoun commented on launch/start.launch:
```diff
@@ -31,6 +31,7 @@
   <arg name="boresight_calibration_file_path" default="$(find polhemus_ros_driver)/config"/>
   <arg name="boresight_calibration_file_name" default="$(arg product_type)_calibration.yaml"/>
   <arg name="boresight_calibration_file" default="$(arg boresight_calibration_file_path)/$(arg boresight_calibration_file_name)"/>
+  <arg name="calibration_data_divisor" default="1"/>
```
Do you want to keep this as 1?

jeferal commented:
I reviewed and would accept, but first perhaps it would be useful to check how the calibration quality changes when changing the tf_data_divisor as we have discussed

carebare47 commented:
> I reviewed and would accept, but first perhaps it would be useful to check how the calibration quality changes when changing the tf_data_divisor as we have discussed

Agreed, checking that now

Commit:
turns out '_load_tf_callback' was missing ~80% of messages due to it's execution time being too long (the number of msgs recieved was system-dependant and made it impossible to compare accuracy of calibration fits between runs). Modified code to save all tf messsages and process them in get_knuckle_positions. Added code to load a bag file for testing/tuning/comparing accuracy of calibration fits. Now that _load_tf_callback is recieving all the messages, the average number of msgs per second has gone up from ~15 to ~100, drastically increasing the computation of sphere fit. But, now we can add code to downsample to a KNOWN number of 'datapoints per second', giving the same performance across systems and lettings us actually figure out how many datapoints per second we need for a reasonable calibration fit

Commit:
now saves all tf messages and processes outside the callback, downsampling to approximately a 'known' number of samples-per-second. Switched to MarkerArrays for faster marker publishing. Also general tidying

Commit:
make sure  is valid for large values of

carebare47 commented:
> I reviewed and would accept, but first perhaps it would be useful to check how the calibration quality changes when changing the tf_data_divisor as we have discussed

It turns out that the `/tf` subscriber was missing the vast majority of messages because the callback was taking too long to execute, so it wasn't possible to compare/find a good "data divisor" number because we couldn't be sure that we were receiving the same messages every time (even from a bag file).

Added code to plug in a bag file of glove movements for testing purposes and moved the processing out of the callback so that the code catches every `/tf` message (either live or from a bag). Now, instead of dividing an unknown amount of incoming data, we capture all the data and subsample to give us approximately `desired_number_of_datapoints_per_sec`. Should be consistent across different machines now.

The bag file I've been testing with has ~120 messages per second.

Setting `desired_number_of_datapoints_per_sec` to above 120:
* Calibration score of `96.2%`
* Sum of time taken to process tf data, publish markers and calculate the sphere fits: ~19 seconds

Setting `desired_number_of_datapoints_per_sec` to `20`:
* Calibration score of `96.3%`
* Sum of time taken to process tf data, publish markers and calculate the sphere fits: ~2.5 seconds

Setting `desired_number_of_datapoints_per_sec` to `10`:
* Calibration score of `96.2%`
* Sum of time taken to process tf data, publish markers and calculate the sphere fits: ~1.5 seconds

(Tests performed on a Ryzen 9 3900X)

So, we should tune this value when we test on an events laptop in spectrum house soon

Commit:
add comment to explain

carebare47 commented:
while I remember, must update the rviz configs in sr_glove_calibration_gui before merging this. Will do this during testing

Commit:
now set timer rate to desired_num_datapoints_per_sec for testing mode with a bag

bdenoun commented on src/polhemus_ros_driver/knuckle_position_calibration_action_server.py:
```diff
@@ -96,17 +96,49 @@ def __init__(self, _hand_prefix: str):
+        self._marker_array.markers.append(data_point_marker)
+
+    def get_number_of_markers(self):
+        return self._marker_array.markers.__len__()
```
Although valid, why not 
```suggestion
        return len(self._marker_array.markers)
```
? Is the __len__ method of marker_array does anything different than the `len()` operator on iterable?

bdenoun commented on src/polhemus_ros_driver/knuckle_position_calibration_action_server.py:
```diff
@@ -96,17 +96,49 @@ def __init__(self, _hand_prefix: str):
-    def __init__(self, side: str = "right"):
+    def __init__(self, side: str = "right",
+                 desired_datapoints_per_sec: int = 20,
+                 testing_bag_file_path: str = ''):
```
Usually, when an argument is optional, we set it to `None`, because an empty string remains a string, which is the valid "type" of data.
Is there a reason you used the empty string?
If you want to change this to None, then this becomes
```suggestion
                 testing_bag_file_path: Optional[str] = None) -> None:
```
while adding `from typing import Optional`

bdenoun replied:
```suggestion
                 testing_bag_file_path: str = '') -> None:
```
to add the output type hint

bdenoun commented on src/polhemus_ros_driver/knuckle_position_calibration_action_server.py:
```diff
@@ -73,10 +73,10 @@ class DataMarker(Marker):
     _id = 0
 
-    def __init__(self, frame_id: str, point: Point, color: ColorRGBA, size=0.002):
+    def __init__(self, frame_id: str, time_stamp: rospy.Time, point: Point, color: ColorRGBA, size=0.002):
```
Adding output type hint
```suggestion
    def __init__(self, frame_id: str, time_stamp: rospy.Time, point: Point, color: ColorRGBA, size=0.002) -> None:
```

bdenoun commented on src/polhemus_ros_driver/knuckle_position_calibration_action_server.py:
```diff
@@ -96,17 +96,49 @@ def __init__(self, _hand_prefix: str):
+        self._marker_array = MarkerArray()
+        self._pub = rospy.Publisher(f"/data_point_marker_{self.hand_prefix}", MarkerArray, queue_size=1000)
+
+    def add_marker(self, frame_id: str, time_stamp: rospy.Time, point: Point, color: ColorRGBA):
```
```suggestion
    def add_marker(self, frame_id: str, time_stamp: rospy.Time, point: Point, color: ColorRGBA) -> None:
```

bdenoun commented on src/polhemus_ros_driver/knuckle_position_calibration_action_server.py:
```diff
@@ -96,17 +96,49 @@ def __init__(self, _hand_prefix: str):
+        data_point_marker = DataMarker(frame_id=frame_id, time_stamp=time_stamp, point=point, color=color)
+        self._marker_array.markers.append(data_point_marker)
+
+    def get_number_of_markers(self):
```
```suggestion
    def get_number_of_markers(self) -> int:
```

bdenoun commented on src/polhemus_ros_driver/knuckle_position_calibration_action_server.py:
```diff
@@ -96,17 +96,49 @@ def __init__(self, _hand_prefix: str):
+    def get_number_of_markers(self):
+        return self._marker_array.markers.__len__()
+
+    def publish_markers(self):
```
```suggestion
    def publish_markers(self) -> None:
```

bdenoun commented on src/polhemus_ros_driver/knuckle_position_calibration_action_server.py:
```diff
@@ -96,17 +96,49 @@ def __init__(self, _hand_prefix: str):
+        # Empty the marker array after publishing
+        self._marker_array = MarkerArray()
+
+    def remove_published_markers(self):
```
```suggestion
    def remove_published_markers(self) -> None:
```

bdenoun commented on src/polhemus_ros_driver/knuckle_position_calibration_action_server.py:
```diff
@@ -140,6 +172,13 @@ def __init__(self, side: str = "right"):
+        self._desired_datapoints_per_sec = desired_datapoints_per_sec
         self._action_server.start()
 
     def _update_current_knuckle_tf(self, hand: Hand):
```
```suggestion
    def _update_current_knuckle_tf(self, hand: Hand) -> None:
```

bdenoun commented on src/polhemus_ros_driver/knuckle_position_calibration_action_server.py:
```diff
@@ -390,20 +431,53 @@ def _create_control(quaternion: Quaternion, name: str):
-                        data_point_marker = DataMarker(hand.polhemus_base_name, Point(*pos),
-                                                       COLORS[color_index].value)
-                        hand.pub.publish(data_point_marker)
+    def _tf_callback(self, data):
```
missing input and output type hint

bdenoun commented on src/polhemus_ros_driver/knuckle_position_calibration_action_server.py:
```diff
@@ -390,20 +431,53 @@ def _create_control(quaternion: Quaternion, name: str):
+            self._saved_tf_msgs.append(data)
+            return
+
+    def _process_tf_data(self, time_elapsed):
```
Missing type hint

bdenoun commented on src/polhemus_ros_driver/knuckle_position_calibration_action_server.py:
```diff
@@ -390,20 +431,53 @@ def _create_control(quaternion: Quaternion, name: str):
+        # Subsample the saved_tf_msgs (temporally) to approximately the desired number of datapoints per second
+        num_msgs = len(self._saved_tf_msgs)
+        desired_num_msgs = time_elapsed * self._desired_datapoints_per_sec
+        subsampled_tf_msgs = self._saved_tf_msgs[::int(round((num_msgs / desired_num_msgs)+0.5))]
```
I think a small comment re. the operation used after `::` would be useful to the future engineers who might need to maintain this script

bdenoun commented on src/polhemus_ros_driver/knuckle_position_calibration_action_server.py:
```diff
@@ -390,20 +431,53 @@ def _create_control(quaternion: Quaternion, name: str):
+                for hand in self._hands.values():
+                    for color_index, finger in enumerate(fingers):
+                        if individual_transform.child_frame_id == hand.finger_data[finger]['polhemus_tf_name']:
+                            pos = [individual_transform.transform.translation.x,
```
```suggestion
                            position = [individual_transform.transform.translation.x,
```

bdenoun commented on src/polhemus_ros_driver/knuckle_position_calibration_action_server.py:
```diff
@@ -390,20 +431,53 @@ def _create_control(quaternion: Quaternion, name: str):
+                                   individual_transform.transform.translation.y,
+                                   individual_transform.transform.translation.z]
+
+                            hand.finger_data[finger]['data'].append(pos)
```
```suggestion
                            hand.finger_data[finger]['data'].append(position)
```

bdenoun commented on src/polhemus_ros_driver/knuckle_position_calibration_action_server.py:
```diff
@@ -390,20 +431,53 @@ def _create_control(quaternion: Quaternion, name: str):
+                            hand.finger_data[finger]['data'].append(pos)
+                            hand.add_marker(frame_id=hand.polhemus_base_name,
+                                            time_stamp=individual_transform.header.stamp,
+                                            point=Point(*pos),
```
```suggestion
                                            point=Point(*position),
```

bdenoun commented on src/polhemus_ros_driver/knuckle_position_calibration_action_server.py:
```diff
@@ -390,20 +431,53 @@ def _create_control(quaternion: Quaternion, name: str):
+        for hand in self._hands.values():
+            hand.publish_markers()
+
+    def _unpack_bag_msg_to_tf_callback(self, _event=False):
```
Missing type hint

bdenoun commented on src/polhemus_ros_driver/knuckle_position_calibration_action_server.py:
```diff
@@ -390,20 +431,53 @@ def _create_control(quaternion: Quaternion, name: str):
+            Grab the next message from the filtered generator and pass it to the tf_callback
+        """
+        try:
+            topic_msg_t = next(self._bag_msgs_generator)
```
What does `_t` mean? Time? could be made clearer?

bdenoun commented on src/polhemus_ros_driver/knuckle_position_calibration_action_server.py:
```diff
@@ -422,27 +496,50 @@ def _calibration(self, goal: CalibrateGoal):
         _result = CalibrateResult()
 
-        sub = rospy.Subscriber("/tf", TFMessage, self._load_tf_callback, queue_size=10)
+        if self._testing_bag_file_path == '':
```
As previously mentioned, would replace this by 
```suggestion
        if self._testing_bag_file_path is None:
```

bdenoun commented on src/polhemus_ros_driver/knuckle_position_calibration_action_server.py:
```diff
@@ -140,6 +172,13 @@ def __init__(self, side: str = "right"):
+        self._progress_period = 1.0 / self.NUMBER_OF_CHECKPOINTS
+        # To store tf messages so we can process them outside the callback
+        self._saved_tf_msgs = []
+        self._testing_bag_file_path = testing_bag_file_path
```
You might want to ensure that the input file (if not None) points to a valid path

bdenoun commented on src/polhemus_ros_driver/knuckle_position_calibration_action_server.py:
```diff
@@ -422,27 +496,50 @@ def _calibration(self, goal: CalibrateGoal):
+        else:
+            # Use /tf data from a bag file (for testing)
+            self._bag_msgs_generator = (topic_msg_t for topic_msg_t in
+                                        rosbag.Bag(self._testing_bag_file_path).read_messages()
```
as previously mention I think that before this operation, a test on the path validity should be made

bdenoun commented on src/polhemus_ros_driver/knuckle_position_calibration_action_server.py:
```diff
@@ -422,27 +496,50 @@ def _calibration(self, goal: CalibrateGoal):
 
-        sub.unregister()
-        self._get_knuckle_positions(hand)
+        if self._testing_bag_file_path == '':
```
```suggestion
        if self._testing_bag_file_path is None':
```

bdenoun commented on src/polhemus_ros_driver/knuckle_position_calibration_action_server.py:
```diff
@@ -466,30 +563,43 @@ def _remove_all_markers(hand: Hand):
+        hand.remove_published_markers()
 
-    def _get_knuckle_positions(self, hand: Hand, plot=False):
+    def _get_knuckle_positions(self, hand: Hand, time_elapsed: float, plot=False):
```
Missing output type hint. I know it's not your fault, but this naming convention of `_get_x` that does not return anything (i.e. not a getter) confuses me a little... I think it would be better called `_compute_knuckle_positions`. What do you think?

bdenoun commented on src/polhemus_ros_driver/knuckle_position_calibration_action_server.py:
```diff
@@ -466,30 +563,43 @@ def _remove_all_markers(hand: Hand):
         """
             Updates the current solution for all fingers on the selected hand.
             @param hand: Selected hand
+            @param time_elapsed: Time elapsed since the start of this calibration
```
Can you describe what's the unit? seconds? milliseconds?

bdenoun commented on src/polhemus_ros_driver/knuckle_position_calibration_action_server.py:
```diff
@@ -509,11 +619,16 @@ def get_calibration_quality(hand: Hand):
         quality_list: List[float] = []
         for finger in fingers:
+            if len(hand.finger_data[finger]['residual']) == 0:
+                print('no data recieved')
```
```suggestion
                rospy.loginfo(f'No data received for finger {finger}')
```

bdenoun commented on src/polhemus_ros_driver/knuckle_position_calibration_action_server.py:
```diff
@@ -509,11 +619,16 @@ def get_calibration_quality(hand: Hand):
         for finger in fingers:
+            if len(hand.finger_data[finger]['residual']) == 0:
+                print('no data recieved')
             quality_list.append(np.std(hand.finger_data[finger]['residual']))
```
We still add the standard deviation of a non-existing distribution?

Commit:
update default num datapoints per sec to 60

Commit:
fix difference between bags and live tf messages, address PR comments

Commit:
Merge branch 'F_#HE_948_improve_calibration_process' of github.com:shadow-robot/polhemus_ros_driver into F_#HE_948_improve_calibration_process

Commit:
oops

Commit:
lint

carebare47 committed:
Merge branch 'noetic-devel' into F_#HE_948_improve_calibration_process

carebare47 committed:
Merge branch 'noetic-devel' into F_#HE_948_improve_calibration_process

Commit:
fixing cal score calc

Commit:
lint
